### PR TITLE
Statusbar Color

### LINKED ISSUE #85: [Proposal] StatusBarEffect
# StatusBarEffect

* [x] Proposed
* [x] Prototype
* [x] Implementation
  * [x] iOS Support
  * [x] Android Support
  * [ ] macOS Support
  * [ ] Windows Support
* [x] Unit Tests: Not Started
* [x] Sample
* [x] Documentation

## Summary
[summary]: #summary

Set the color and style of the status bar

## Detailed Design
[design]: #detailed-design

### StatusBarEffect.shared.cs

```cs
public class StatusBarEffect : RoutingEffect
{
  public static readonly BindableProperty ColorProperty;
  public static readonly BindableProperty StyleProperty;
  
  public static Color GetColor(BindableObject bindable);
  public static StatusBarStyle GetStyle(BindableObject bindable);
}
```

## Usage Syntax
[usage]: #usage-syntax

### XAML Usage

```xml
<ContentPage local:StatusBarEffect.Color="Blue">

</ContentPage>
```

### C# Usage

```cs
class MyPage : ContentPage
{
  public MyPage()
  {
     StatusBarEffect.SetColor(this, Colors.Black);
  }
}
```

### PR DESCRIPTION
**Pull request by pictos**
### Description of Change ###

Adds this amazing feature to this amazing project.

- This project adds a new `namespace` in the `Core` project, called `Capabilities` (I'm fine to change the name, I'm terrible at naming things). The idea is to add in this namespace all APIs that access some platform-specific capability (like Essential does)

 ### Linked Issues ###
 

 - Fixes #85 

 ### PR Checklist ###
 
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [ ] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/XYZ 

 ### Additional information ###

~As I'll post in the original issue, this PR just adds the feature to change the Color of the statusbar. I drop the `StatusBarStyle` because apple changed the way to customize that and the current way to do is overriding the `preferredStatusBarStyle` and we can't do that here... The change should happen in .NET MAUI itself providing a way to control it.~

Thanks to @KSemenenko for the help and code snippet. Now we have statusBarStyle implemented❣

I love this community <3

### CODE REVIEW

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs
Comment (brminnick) [inline src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs:14]: I've never used `window.ViewWithTag()` before.

Where is tag #38482 set? Is that set in OS by Apple or in .NET MAUI? Any fears of this number changing in the future by Apple or .NET MAUI?

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.windows.cs
Comment (brminnick) [inline src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.windows.cs:9]: Should this be `NotSupportedException`?

Comment (brminnick) [inline src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.windows.cs:9]:
```suggestion
		throw new NotSupportedException("WinUI does not currently support changing the Windows status bar color");
```

File: src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/StatusBar/StatusBarBehavior.shared.cs
Comment (brminnick) [inline src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/StatusBar/StatusBarBehavior.shared.cs:13]: Should this be `PlatformBehavior<Page>`? 

For example, I'm not sure I can think of a scenario where we would attach a `StatusBarBehavior` to a `Label`.

```suggestion
public class StatusBarBehavior : PlatformBehavior<Page>
```

**Review by brminnick — COMMENTED** [review]
Thanks Pedro! 

Agreed - we should use a more specific name than `Capabilities`...but I also can't think of what we should use instead 😆

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.windows.cs
Comment (brminnick) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.windows.cs:9]:
```suggestion
		throw new NotSupportedException("WinUI does not currently support changing the Windows status bar color");
```

File: src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/StatusBar/StatusBarBehavior.shared.cs
Comment (pictos) [reply src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/StatusBar/StatusBarBehavior.shared.cs:13]: Yeah, I used the `VisualElement` to make it more general, so devs can apply the behavior everywhere. But I'm ok to make it more restrict

**Review by brminnick — CHANGES_REQUESTED** [review]
Don't forget the Unit Tests! 🙂

**Comment by pictos** [comment]
@brminnick this is a `PlatformBehavior` we can't test it in `net6` world, we need the device-runner.

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs
Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs:14]: Still looking for this magic number, but didn't find the explanation yet... But looks like the people on [swift world use](https://mobikul.com/get-status-bar-view-in-ios-13-and-above/) this number as well...
Maybe @KSemenenko knows the history behind this magic number...

File: src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/StatusBar/StatusBarBehavior.shared.cs
Comment (pictos) [reply src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/StatusBar/StatusBarBehavior.shared.cs:13]: So, will we change this to be `Page` or let it be as it is?

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs
Comment (KSemenenko) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs:14]: It works by adding a box or something like that.
And this id everyone uses as a marker that this is the color of the status bar :)

File: src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/StatusBar/StatusBarBehavior.shared.cs
Comment (KSemenenko) [reply src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/StatusBar/StatusBarBehavior.shared.cs:13]: The page is the main consumer.
But maybe you want to make the status bar red to show something?
Like an error. Or that such an app is doing something.

Comment (KSemenenko) [reply src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/StatusBar/StatusBarBehavior.shared.cs:13]: The problem is that this code is not tied to the page. It's global for the entire application.
And when we go to each page, we have to use this code to get the status bar of the desired color.
 Otherwise the style from previous page will remain.

File: src/CommunityToolkit.Maui.Core/Capabilities/PlatformVersion/PlatformVersion.android.cs
Comment (maxkoshevoi) [inline src/CommunityToolkit.Maui.Core/Capabilities/PlatformVersion/PlatformVersion.android.cs:15]: Cannot `DeviceInfo.Version` be used for this? This creates another source of truth for getting device version

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs
Comment (maxkoshevoi) [inline src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs:42]: nit: Remove one space
```suggestion
		SetBarAppearance(activity, barAppearance => barAppearance |= flag);
```

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs
Comment (maxkoshevoi) [inline src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs:52]: `SystemUiVisibility` is deprecated as of API 30 (Android 11). Have you tested this code on Android 11-13 devices?
This worked in XCT since it targeted lover version of Android.

Here're needed changes to support Android 11+: [xamarin/XamarinCommunityToolkit@`38e872a` (#812)](https://github.com/xamarin/XamarinCommunityToolkit/pull/812/commits/38e872a1c54c4956ef0bc62edbb423059eb5d426)

Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs:52]: I tested it on Android 12 and it worked. We (Core team members) know this is deprecated, but we want to release this feature ASAP, so we will use the obsolete version for now and implement the new approach soon. I suppose that you want to contribute to that, if so, I'll make sure to ping and assign you to this task when it's created.

Comment (brminnick) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs:52]: @pictos - could you open an Issue for the use of this deprecated before we merge this PR? That way we won't forget to fix it in the future!

File: src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/StatusBar/StatusBarBehavior.shared.cs
Comment (pictos) [reply src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/StatusBar/StatusBarBehavior.shared.cs:13]: @KSemenenko for now we will let the responsibility to the user, if we see a demand we can implement this later

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs
Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs:52]: @brminnick here's the issue
- https://github.com/CommunityToolkit/Maui/issues/626
@maxkoshevoi if you want to work on that, please add a comment on the issue until Friday.

Comment (maxkoshevoi) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs:52]: Hm, strange how it works while being deprecated. I'm sure it was crashing on Android 11 device when I retargeted XCT to Android 11, and used deprecated code.
Thank you for creating an issue, added a message there.

File: src/CommunityToolkit.Maui.Core/Capabilities/PlatformVersion/PlatformVersion.android.cs
Comment (maxkoshevoi) [reply src/CommunityToolkit.Maui.Core/Capabilities/PlatformVersion/PlatformVersion.android.cs:16]: @pictos Why is this resolved? I don't see any related commit. Could you at least comment on why this code is nesesary? (It was needed in XCT since Essentials were in the separate NuGet)

Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Capabilities/PlatformVersion/PlatformVersion.android.cs:16]: This code is more performatic than Maui ones. The core project is meant to be used by other toolkit implementors and not by app projects, so devs should not use it.
I added one more line in the `summary` telling this code is for internal use only.

**Comment by brminnick** [comment]
Hey Gang! Just checking in - are we good to merge this PR today or tomorrow (Wednesday or Thursday)?

We plan on pushing a new release to NuGet on Friday and would love to include this PR! 

If it needs more time, no worries! We can add it in the next release 💯

**Comment by pictos** [comment]
@brminnick this is ready for the last code review, when it's approved I'll write the docs

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs:73]: Can we store it in some field to not calculate it each time? We do the same in DrawingView iOS handler

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.net.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.net.cs:5]: Maybe we should log a message that it is not supported, like you have in Android (IsSupported method). It may be a blocker for devs who want to release on Windows.

or use SuppressException options in Core.Options

File: src/CommunityToolkit.Maui.Core/Capabilities/PlatformVersion/PlatformVersion.android.cs
Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Capabilities/PlatformVersion/PlatformVersion.android.cs:9]: If other toolkits want to use it they can. I'm ok adding the attribute to hide it from the intellisense. What do you think?

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs
Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs:69]: On Android doesn't make sense, because the APIs are different.

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs
Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs:73]: store what?

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.net.cs
Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.net.cs:5]: I added the attribute telling this is not supported on windows. I'm pretty sure I did it for the Behavior, not sure for the Class in core, need to double check that

File: src/CommunityToolkit.Maui.Core/Capabilities/PlatformVersion/PlatformVersion.android.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Capabilities/PlatformVersion/PlatformVersion.android.cs:9]: I don’t see the reason to add comment that it’s for internal usage. If it’s for internal usage it should be internal.
I would just remove it from summary.

maybe add comment that it’s an alternative to.NET MAUI implementation

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs:69]: Can we have a method/property isAndroidMOrNewer?

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs:73]: VC

File: src/CommunityToolkit.Maui.Core/Capabilities/PlatformVersion/PlatformVersion.android.cs
Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Capabilities/PlatformVersion/PlatformVersion.android.cs:9]: Just follow the pattern used by.NET MAUI and XF.

image

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.net.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.net.cs:5]: Yes, it will require devs to add #if !Windows which may be annoying.
Debug.WriteLine() is more friendly. Anyway we add in docs that it is not supported on windows

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs
Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs:69]: I mean, why do we want to add more complexity? IMHO, since this code is just in Android world (`.android` file) doesn't add value trying to make it similar to iOS. They are two different worlds.

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.net.cs
Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.net.cs:5]: `Debug.WriteLine` will print nothing in the package :/

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs
Comment (pictos) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs:73]: Not sure if the app uses multi-window.

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.android.cs:69]: If we use it only in 1 place do we need PlatformVersion class at all?

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.ios.cs:73]: It makes sense, resolved

File: src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.net.cs
Comment (KSemenenko) [reply src/CommunityToolkit.Maui.Core/Capabilities/StatusBar/StatusBar.net.cs:5]: Debug.WriteLine is terrible :)

File: src/CommunityToolkit.Maui.Core/Platform/StatusBar/StatusBar.macos.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Platform/StatusBar/StatusBar.macos.cs:6]: I don't think we support macOS

File: src/CommunityToolkit.Maui.Core/Platform/StatusBar/StatusBar.net.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Platform/StatusBar/StatusBar.net.cs:5]: looks like the message is not correct. we do not support maccatalyst and windows

File: src/CommunityToolkit.Maui.Core/Platform/StatusBar/StatusBar.android.cs
Comment (brminnick) [reply src/CommunityToolkit.Maui.Core/Platform/StatusBar/StatusBar.android.cs:71]: Yup! The attribute `[UnsupportedOSPlatform]` just generates a compiler warning. We need this to ensure the operating system version is verified at runtime

File: src/CommunityToolkit.Maui.Core/Platform/StatusBar/StatusBar.macos.cs
Comment (brminnick) [reply src/CommunityToolkit.Maui.Core/Platform/StatusBar/StatusBar.macos.cs:6]: Correct - we don't.

I did this just because I was unsure how robust/smart the `UnsupportedOSPlatform` is. 

I.e., I was unsure if would it would generate a compiler warning if only `UnsupportedOSPlatform("MacCatalyst")` was listed.

File: src/CommunityToolkit.Maui.Core/Platform/StatusBar/StatusBar.net.cs
Comment (brminnick) [reply src/CommunityToolkit.Maui.Core/Platform/StatusBar/StatusBar.net.cs]: Yup! Good catch! Fixed ✅

File: src/CommunityToolkit.Maui.Core/Platform/StatusBar/StatusBar.macos.cs
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.Core/Platform/StatusBar/StatusBar.macos.cs:6]: our target framework is maccatalyst, so MacOS is not necessary

**Review by VladislavAntonyuk — APPROVED** [review]

**Review by brminnick — DISMISSED** [review]
Thanks Pedro! I LOVE the `StatusBarBehaviorPage` in the sample app!! It's such an awesome + easy demonstration of how to use the `StatusBarBehavior` 💯

I updated the base class of `StatusBarBehavior` to inherit from `PlatformBehavior<Page>`, since it doesn't make much sense for devs to attach this behavior to something like a `Label`.

I renamed `Core.Capabilities` to `Core.Platform`. This better aligns to `Microsoft.Maui.Platform` and should make it easier to (hopefully!) promote this feature into .NET MAUI in the future! 

I removed `PlatformVersion` since .NET 6.0 already provides this functionality via `OperatingSystem.IsIOSVersionAtLeast()` and `OperatingSystem.IsAndroidVersionAtLeast()`. Plus, removing it means we have one less API to maintain + unit test.

I also made a couple syntax tweaks to the Sample App, consolidating some code in `StatusBarBehaviorViewModel`, and centering the `RadioButton` Text, adding some margins for spacing and centering the text.

**Review by brminnick — APPROVED** [review]
Approving again after running `dotnet format`

**Comment by brminnick** [comment]
@pictos Are you able to knock out the Docs quickly for this PR? 

I'm going to bend protocol a bit and merge this before the docs have been written + approved in order to squeeze this in for tomorrow's release 👍

**Review by VladislavAntonyuk — APPROVED** [review]

**Comment by samirgcofficial** [comment]
Are there any plans for working on Windows and Mac. My client is specific to Windows and Mac dotnet maui apps. ?

**Comment by bijington** [comment]
> Are there any plans for working on Windows and Mac. My client is specific to Windows and Mac dotnet maui apps. ?

From what I understand we are limited by the underlying layers (WinUI and MacCatalyst) not providing support for this. So to answer your question, yes there are plans but only once the platforms allow us

**Pull request merged**
